Userland/Libraries/LibTLS/TLSv12.cpp

```diff
@@ -4,622 +4,122 @@
+    static char buffer[WOLFSSL_MAX_ERROR_SZ];
+    auto* ptr = wolfSSL_ERR_error_string(error, buffer);
+    return StringView { ptr, strlen(ptr) };
 }
 
-static bool certificate_subject_matches_host(Certificate const& cert, StringView host)
+ErrorOr<NonnullOwnPtr<WolfTLS>> WolfTLS::connect(const AK::ByteString& host, u16 port)
 {
-    if (wildcard_matches(host, cert.subject.common_name()))
-        return true;
-
-    for (auto& san : cert.SAN) {
-        if (wildcard_matches(host, san))
-            return true;
+    if (!g_initialized_wolfssl) {
+        wolfSSL_Init();
+        g_initialized_wolfssl = true;
     }
 
-    return false;
-}
+    auto* context = wolfSSL_CTX_new(wolfTLSv1_2_client_method());
+    if (!context)
+        return Error::from_string_literal("Failed to create a new TLS context");
 
-bool Context::verify_chain(StringView host) const
-{
-    if (!options.validate_certificates)
-        return true;
+    if (s_certificate_store_paths.is_empty())
+        s_certificate_store_paths.append("/etc/ssl/cert.pem"); // We're just guessing this, the embedder should provide this.
 
-    Vector<Certificate> const* local_chain = nullptr;
-    if (is_server) {
-        dbgln("Unsupported: Server mode");
-        TODO();
-    } else {
-        local_chain = &certificates;
+    for (auto& path : s_certificate_store_paths) {
+        if (wolfSSL_CTX_load_verify_locations(context, path.characters(), nullptr) != SSL_SUCCESS)
+            return Error::from_string_literal("Failed to load CA certificates");
     }
 
-    if (local_chain->is_empty()) {
-        dbgln("verify_chain: Attempting to verify an empty chain");
-        return false;
-    }
+    auto ssl = wolfSSL_new(context);
+    if (!ssl)
+        return Error::from_string_literal("Failed to create a new SSL object");
 
-    // RFC5246 section 7.4.2: The sender's certificate MUST come first in the list. Each following certificate
-    // MUST directly certify the one preceding it. Because certificate validation requires that root keys be
-    // distributed independently, the self-signed certificate that specifies the root certificate authority MAY be
-    // omitted from the chain, under the assumption that the remote end must already possess it in order to validate
-    // it in any case.
+    auto tcp_socket = TRY(Core::TCPSocket::connect(host, port));
+    auto fd = tcp_socket->fd();
 
-    if (!host.is_empty()) {
-        auto const& first_certificate = local_chain->first();
-        auto subject_matches = certificate_subject_matches_host(first_certificate, host);
-        if (!subject_matches) {
-            dbgln("verify_chain: First certificate does not match the hostname");
-            return false;
-        }
-    } else {
-        // FIXME: The host is taken from m_context.extensions.SNI, when is this empty?
-        dbgln("FIXME: verify_chain called without host");
-        return false;
+    if (wolfSSL_set_fd(ssl, fd) != SSL_SUCCESS) {
```

Comment:
TODO: Use https://www.wolfssl.com/documentation/manuals/wolfssl/wolfio_8h.html instead of hacking TCPSocket.